src/yunohost/backup.py

```diff
@@ -483,83 +709,286 @@ def _is_hook_in_backup(h):
             apps_filtered = apps_list
 
         for app_instance_name in apps_filtered:
-            tmp_app_dir = '{:s}/apps/{:s}'.format(tmp_dir, app_instance_name)
-            tmp_app_bkp_dir = tmp_app_dir + '/backup'
+            self._restore_app(app_instance_name)
+
+    def _restore_app(self, app_instance_name):
+        def copytree(src, dst, symlinks=False, ignore=None):
+            for item in os.listdir(src):
+                s = os.path.join(src, item)
+                d = os.path.join(dst, item)
+                if os.path.isdir(s):
+                    shutil.copytree(s, d, symlinks, ignore)
+                else:
+                    shutil.copy2(s, d)
+
+        tmp_app_dir = os.path.join(self.mnt_path, 'apps', app_instance_name)
+        tmp_app_bkp_dir = os.path.join(tmp_app_dir, 'backup')
+
+        # Parse app instance name and id
+        # TODO: Use app_id to check if app is installed?
+        app_id, app_instance_nb = _parse_app_instance_name(app_instance_name)
+
+        # Check if the app is not already installed
+        if _is_installed(app_instance_name):
+            logger.error(m18n.n('restore_already_installed_app',
+                                app=app_instance_name))
+            return
+
+        # Check if the app has a restore script
+        app_script = os.path.join(tmp_app_dir, 'settings/scripts/restore')
+        if not os.path.isfile(app_script):
+            logger.warning(m18n.n('unrestore_app', app=app_instance_name))
+            return
+
+        tmp_settings_dir = os.path.join(tmp_app_dir, 'settings')
+        app_setting_path = os.path.join('/etc/yunohost/apps/',
+                                        app_instance_name)
+        logger.info(m18n.n('restore_running_app_script', app=app_instance_name))
+        try:
+            # Copy scripts to a writable temporary folder
+            tmp_scripts_dir = tempfile.mkdtemp(prefix='restore')
+            copytree(os.path.join(tmp_settings_dir, 'scripts'), tmp_scripts_dir)
+            filesystem.chmod(tmp_scripts_dir, 0550, 0550, True)
+            filesystem.chown(tmp_scripts_dir, 'admin', None, True)
+            app_script = os.path.join(tmp_scripts_dir, 'restore')
+
+            # Copy app settings and set permissions
+            # TODO: Copy app hooks too
+            shutil.copytree(tmp_settings_dir, app_setting_path)
+            filesystem.chmod(app_setting_path, 0400, 0400, True)
+            filesystem.chown(os.path.join(app_setting_path, 'scripts'),
+                             'admin', None, True)
+
+            # Prepare env. var. to pass to script
+            env_dict = {}
+            env_dict["YNH_APP_ID"] = app_id
+            env_dict["YNH_APP_INSTANCE_NAME"] = app_instance_name
+            env_dict["YNH_APP_INSTANCE_NUMBER"] = str(app_instance_nb)
+            env_dict["YNH_APP_BACKUP_DIR"] = tmp_app_bkp_dir
```

Comment:
I think we need YNH_BACKUP_CSV also (ynh_restore_file complains about it missing)